### PR TITLE
VW: Fix permission checking in `root`

### PR DESCRIPTION
**Pull request by davidfestal**
## Summary

We should only skip permission checking in the `root` workspace for the `list` verb, so that any user can always list the workspaces he has `access` to and can `get`.

## Related issue(s)

No issues

### CODE REVIEW

File: pkg/virtual/workspaces/registry/rest.go [hunk 1]
```diff
@@ -269,9 +269,12 @@ func (s *REST) deprecatedAuthorizeOrgForUser(ctx context.Context, orgClusterName
...
 func (s *REST) authorizeForUser(ctx context.Context, orgClusterName logicalcluster.Name, currentUser kuser.Info, verb string, resourceName string) error {
-	// Root org access is implicit for every user. For non-root orgs, we need to check for
-	// verb permissions against the clusterworkspaces/workspace sub-resource.
-	if orgClusterName == tenancyv1alpha1.RootCluster || sets.NewString(currentUser.GetGroups()...).Has(kuser.SystemPrivilegedGroup) {
+	if sets.NewString(currentUser.GetGroups()...).Has(kuser.SystemPrivilegedGroup) {
+		return nil
+	}
+
+	// Root org workspace list is implicit for every user.
+	if orgClusterName == tenancyv1alpha1.RootCluster && verb == "list" {
```
Comment (sttts) [inline pkg/virtual/workspaces/registry/rest.go:277]: why not `get` too?

File: pkg/virtual/workspaces/registry/rest.go [hunk 2]
```diff
@@ -269,32 +269,31 @@ func (s *REST) deprecatedAuthorizeOrgForUser(ctx context.Context, orgClusterName
...
-	if parent, _ := orgClusterName.Split(); parent == tenancyv1alpha1.RootCluster {
-		switch verb {
-		case "get", "watch", "list":
-			return s.deprecatedAuthorizeOrgForUser(ctx, orgClusterName, currentUser, "access")
-		case "create":
-			return s.deprecatedAuthorizeOrgForUser(ctx, orgClusterName, currentUser, "member")
-		case "delete":
-			if err := s.deprecatedAuthorizeOrgForUser(ctx, orgClusterName, currentUser, "access"); err != nil {
-				return err
+	if orgClusterName != tenancyv1alpha1.RootCluster {
```
Comment (p0lyn0mial) [inline pkg/virtual/workspaces/registry/rest.go:276]: what if `orgClusterName` == `tenancyv1alpha1.RootCluster` ?

File: pkg/virtual/workspaces/registry/rest.go [hunk 3]
```diff
@@ -269,32 +269,31 @@ func (s *REST) deprecatedAuthorizeOrgForUser(ctx context.Context, orgClusterName
 }
 
 func (s *REST) authorizeForUser(ctx context.Context, orgClusterName logicalcluster.Name, currentUser kuser.Info, verb string, resourceName string) error {
-	// Root org access is implicit for every user. For non-root orgs, we need to check for
-	// verb permissions against the clusterworkspaces/workspace sub-resource.
-	if orgClusterName == tenancyv1alpha1.RootCluster || sets.NewString(currentUser.GetGroups()...).Has(kuser.SystemPrivilegedGroup) {
+	if sets.NewString(currentUser.GetGroups()...).Has(kuser.SystemPrivilegedGroup) {
```
Comment (p0lyn0mial) [inline pkg/virtual/workspaces/registry/rest.go:272]: okay, privileged users always have power

File: pkg/virtual/workspaces/registry/rest.go (hunk 2 above)
Comment (sttts) [reply pkg/virtual/workspaces/registry/rest.go:276]: RootCluster has no parent. Hence we cannot check anything in there. Before this PR we skipped the whole authz step though 😱. Now we only skip the access/member check in the parent.

File: pkg/authorization/bootstrap/policy.go [hunk 4]
```diff
@@ -36,6 +40,19 @@ const (
 func clusterRoleBindings() []rbacv1.ClusterRoleBinding {
 	return []rbacv1.ClusterRoleBinding{
 		clusterRoleBindingCustomName(rbacv1helpers.NewClusterBinding("cluster-admin").Groups(SystemKcpClusterWorkspaceAdminGroup, SystemKcpAdminGroup).BindingOrDie(), "system:kcp:clusterworkspace:admin"),
+		clusterRoleBindingCustomName(rbacv1helpers.NewClusterBinding("system:kcp:tenancy:reader").Groups(SystemKcpClusterWorkspaceAccessGroup).BindingOrDie(), "system:kcp:clusterworkspace:access"),
```
Comment (p0lyn0mial) [inline pkg/authorization/bootstrap/policy.go:43]: so this is saying, `system:kcp:clusterworkspace:access` group will be assigned to `system:kcp:tenancy:reader` role, which allows `list` on `clusterworkspaces/workspace`

Comment (p0lyn0mial) [reply pkg/authorization/bootstrap/policy.go:43]: which users are assigned to `system:kcp:clusterworkspace:access` group?

File: pkg/authorization/bootstrap/policy.go [hunk 5]
```diff
@@ -36,6 +40,19 @@ const (
...
+	}
+}
+
+func clusterRoles() []rbacv1.ClusterRole {
+	return []rbacv1.ClusterRole{
+		{
+			ObjectMeta: metav1.ObjectMeta{Name: "system:kcp:tenancy:reader"},
+			Rules: []rbacv1.PolicyRule{
+				rbacv1helpers.NewRule("list", "watch").Groups(tenancy.GroupName).Resources("clusterworkspaces/workspace").RuleOrDie(), // "get" is by workspace name through workspace VW
```
Comment (p0lyn0mial) [inline pkg/authorization/bootstrap/policy.go:52]: what does `list` mean in this context? (`clusterworkspaces/workspace`)

File: config/root/clusterrolebinding-tenancy-workspace-reader.yaml [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+metadata:
+  name: system:kcp:authenticated:tenancy:reader
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: system:kcp:tenancy:reader
+subjects:
+- apiGroup: rbac.authorization.k8s.io
+  kind: Group
+  name: system:authenticated
```
Comment (p0lyn0mial) [inline config/root/clusterrolebinding-tenancy-workspace-reader.yaml:12]: is this correct?

Comment (p0lyn0mial) [reply config/root/clusterrolebinding-tenancy-workspace-reader.yaml:12]: I think a user can be `system:authenticated` but not in `SystemKcpClusterWorkspaceAccessGroup`

File: pkg/virtual/workspaces/registry/rest.go (hunk 2 above)
Comment (p0lyn0mial) [reply pkg/virtual/workspaces/registry/rest.go:276]: do we have separate checks for users targeting only the `RootCluster`?

Comment (sttts) [reply pkg/virtual/workspaces/registry/rest.go:276]: added the tests.

The logic for root is that everybody has implicit `access` permissions, i.e. can "enter" and has the group membership of `system:kcp:clusterworkspace:access`.

File: pkg/authorization/bootstrap/policy.go (hunk 4 above)
Comment (sttts) [reply pkg/authorization/bootstrap/policy.go:43]: system:kcp:clusterworkspace:access

File: pkg/authorization/bootstrap/policy.go (hunk 5 above)
Comment (sttts) [reply pkg/authorization/bootstrap/policy.go:52]: `kubectl get workspaces`

File: config/root/clusterrolebinding-tenancy-workspace-reader.yaml (hunk 6 above)
Comment (sttts) [reply config/root/clusterrolebinding-tenancy-workspace-reader.yaml:12]: that's intentional. That's one of the batteries in https://github.com/kcp-dev/kcp/pull/1688. In a real service this binding might not make sense. But for demo and development it does.

File: pkg/virtual/workspaces/registry/rest.go (hunk 2 above)
Comment (ncdc) [reply pkg/virtual/workspaces/registry/rest.go:276]: nit: is this `if` statement necessary? The logic inside only executes if orgClusterName's parent is root. That logic works without checking if orgClusterName != root.... right?

File: test/e2e/virtual/workspaces/virtual_workspace_test.go [hunk 7]
```diff
@@ -809,6 +812,122 @@ func testWorkspacesVirtualWorkspaces(t *testing.T, standalone bool) {
...
+				require.NoError(t, err)
+				found := false
+				for _, ws := range wss.Items {
+					if ws.Name == ws2.Name {
+						found = true
+					}
+					require.NotEqual(t, ws.Name, ws1.Name, "user-2 should not see user-1's workspace")
+				}
+				return found, fmt.Sprintf("expected to see workspace %s, got %v", ws2.Name, wss.Items)
+			}, wait.ForeverTestTimeout, time.Millisecond*100, "user1 should see only one workspace")
```
Comment (ncdc) [inline test/e2e/virtual/workspaces/virtual_workspace_test.go:910]:
```suggestion
			}, wait.ForeverTestTimeout, time.Millisecond*100, "user2 should see only one workspace")
```

File: pkg/virtual/workspaces/registry/rest.go [hunk 8]
```diff
@@ -269,32 +269,31 @@ func (s *REST) deprecatedAuthorizeOrgForUser(ctx context.Context, orgClusterName
...
+			switch verb {
+			case "get", "watch", "list":
+				return s.deprecatedAuthorizeOrgForUser(ctx, orgClusterName, currentUser, "access")
+			case "create":
+				return s.deprecatedAuthorizeOrgForUser(ctx, orgClusterName, currentUser, "member")
+			case "delete":
+				if err := s.deprecatedAuthorizeOrgForUser(ctx, orgClusterName, currentUser, "access"); err != nil {
+					return err
+				}
+				// Let's fall through here:
```
[the remark this answers is not in the document]
Comment (sttts) [reply pkg/virtual/workspaces/registry/rest.go:289]: there is. All other cases have a return.

File: pkg/virtual/workspaces/registry/rest.go (hunk 2 above)
Comment (sttts) [reply pkg/virtual/workspaces/registry/rest.go:276]: it's not necessary. True.

File: pkg/virtual/workspaces/registry/rest.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (ncdc) [reply pkg/virtual/workspaces/registry/rest.go:289]: i see... details 😛

**Review by ncdc — APPROVED** [review]

**Pull request merged**